tap_mixpanel/streams.py

```diff
@@ -1,125 +1,654 @@
+            end_window = now_datetime
+            diff_sec = (end_window - start_window).seconds
+            # round-up difference to days
+            days_interval = math.ceil(diff_sec / (3600 * 24))
+
+        return start_window, end_window, days_interval
+
+    def sync(self, state: dict, catalog, config: dict, start_date):
+        # the sync method depending on different endpoints
+
+        bookmark_field = next(iter(self.replication_keys), None)
+        project_timezone = config.get("project_timezone", "UTC")
+        days_interval = int(config.get("date_window_size", "30"))
+        attribution_window = int(config.get("attribution_window", "5"))
+
+        # Get the latest bookmark for the stream and set the last_integer/datetime
+        last_datetime = None
+        max_bookmark_value = None
+        last_datetime = self.get_bookmark(
+            state, self.tap_stream_id, start_date)
+        max_bookmark_value = last_datetime
+
+        self.write_schema(catalog, self.tap_stream_id)
+
+        # windowing: loop through date days_interval date windows from last_datetime to now_datetime
+        tzone = pytz.timezone(project_timezone)
+        now_datetime = datetime.now(tzone)
+        end_date = config.get('end_date')
+
+        if end_date:
+            now_datetime = strptime_to_utc(end_date)
+
+        start_window, end_window, days_interval = self.define_bookmark_filters(
+            days_interval, last_datetime, now_datetime, attribution_window)
+        # LOOP order: Date Windows, Parent IDs, Page
+        # Initialize counter
+        endpoint_total = 0  # Total for ALL: parents, date windows, and pages
 
+        # Begin date windowing loop
+        while start_window < now_datetime:
+            # Initialize counters
+            date_total = 0  # Total records for a date window
+            parent_total = 0  # Total records for parent ID
+            total_records = 0  # Total records for all pages
+            record_count = 0  # Total processed for page
+
+            params = self.params  # adds in endpoint specific, sort, filter params
+
+            if self.bookmark_query_field_from and self.bookmark_query_field_to:
+                # Request dates need to be normalized to project timezone or else errors may occur
+                # Errors occur when from_date is > 365 days ago
+                #   and when to_date > today (in project timezone)
+                from_date = "{}".format(start_window.astimezone(tzone))[0:10]
+                to_date = "{}".format(end_window.astimezone(tzone))[0:10]
+                LOGGER.info(
+                    "START Sync for Stream: {}{}".format(
+                        self.tap_stream_id,
+                        ", Date window from: {} to {}".format(
+                            from_date, to_date)
+                        if self.bookmark_query_field_from
+                        else "",
+                    )
+                )
```

Comment:
This is silly to force into one log message

```suggestion
                LOGGER.info("START Sync for Stream: %s", self.tap_stream_id)
                if self.bookmark_query_field_from:
                    LOGGER.info("Date window from: %s to %s", from_date, to_date)
```

Reply:
Updated